File: 513-find-bottom-left-value/solution.js
/*
BFS看js 自己寫的
DFS看Go 自己寫的
 */
const TreeNode = require('../datastructure/Tree/TreeNode');
const buildLevelOrderTree_v2 = require('../datastructure/Tree/TreeDebugger_v2')

/**
 * @param {TreeNode} root
 * @return {number}
 */
var findBottomLeftValue = function(root) {
    let queue = [root]
    let level = []
    while (queue.length > 0){
        level = []
        let levelSize = queue.length
        for (let i = 0; i < levelSize; i++) {
            let curNode = queue.shift()
            level.push(curNode.val)
            if (curNode.left != null)
                queue.push(curNode.left)
            if (curNode.right != null)
                queue.push(curNode.right)
        }
    }
    return level[0]
};
if (require.main === module){
    const levelOrder = [1, 2, 3, 4, null, 5, 6, null, null, 7]
    const root = buildLevelOrderTree_v2(levelOrder)
    console.log("Ans = ", findBottomLeftValue(root))
}

File: 513-find-bottom-left-value/solution.go
/*
BFS看js
DFS看Go
*/
package main

import (
	"fmt"
	"gopractice/datastructure/Tree"
)

type TreeNode = Tree.TreeNode

func findBottomLeftValue(root *TreeNode) (ans int) {
	ans = root.Val.(int)
	maxDepth := 1
	var dfs func(*TreeNode, int) int
	dfs = func(root *TreeNode, depth int) int {
		if root == nil {
			return 0
		}
		if depth > maxDepth {
			maxDepth = depth
			ans = root.Val.(int)
		}
		dfs(root.Left, depth+1)
		dfs(root.Right, depth+1)
		return 0
	}
	dfs(root, 1)
	return
}
func main() {
	levelOrder := []any{1, 2, 3, 4, nil, 5, 6, nil, nil, 7}
	root := Tree.BuildLevelOrderTree_v2(levelOrder)
	fmt.Println("Ans = ", findBottomLeftValue(root))
}

/*
靈感來自自頂向下的解決方案
https://leetcode.cn/leetbook/read/data-structure-binary-tree/xefb4e/

手寫遞迴樹
dfs(root, 1)
	dfs(root.Left, 2)
		dfs(root.Left, 3)
			dfs(root.Left, 4)(x)
				return 0
				return 0
		dfs(root.Right, 3)(x)
			return 0
	dfs(root.Right, 2)
		dfs(root.Left, 3)
			dfs(root.Left, 4)
				dfs(root.Left, 5)(x)
					return 0
					return 0
			dfs(root.Right, 4)(x)
				return 0
		dfs(root.Right, 3)
			dfs(root.Left, 4)(x)
				return 0
				return 0
*/
